File: _extensions/qant/qant.ts
// qant: Categorical imperative for Quarto categories
// Validates document categories against a centralized allow-list

// Import YAML parser from Deno stdlib
import { parse as parseYaml } from "https://deno.land/std@0.217.0/yaml/mod.ts";

interface InspectData {
  files: {
    [filepath: string]: {
      metadata?: {
        categories?: string[];
      };
    };
  };
}

interface QantConfig {
  categories: string[];
}

interface Violation {
  file: string;
  category: string;
}

function readAllowedCategories(): Set<string> | null {
  try {
    const content = Deno.readTextFileSync("_qant.yml");
    const config = parseYaml(content) as QantConfig;

    if (!config.categories || !Array.isArray(config.categories)) {
      console.warn("Warning: _qant.yml does not contain a valid 'categories' array.");
      return null;
    }

    return new Set(config.categories);
  } catch (error) {
    if (error instanceof Deno.errors.NotFound) {
      console.log("Warning: _qant.yml not found. Skipping category validation.");
      return null;
    }
    throw error;
  }
}

function runQuartoInspect(): InspectData {
  const cmd = new Deno.Command("quarto", {
    args: ["inspect"],
    stdout: "piped",
    stderr: "piped",
  });

  const output = cmd.outputSync();

  if (!output.success) {
    const errorMsg = new TextDecoder().decode(output.stderr);
    throw new Error(`Failed to run quarto inspect: ${errorMsg}`);
  }

  const jsonOutput = new TextDecoder().decode(output.stdout);
  return JSON.parse(jsonOutput) as InspectData;
}

function validateCategories(
  inspectData: InspectData,
  allowedCategories: Set<string>
): Violation[] {
  const violations: Violation[] = [];

  for (const [filepath, filedata] of Object.entries(inspectData.files)) {
    if (filedata.metadata?.categories) {
      for (const category of filedata.metadata.categories) {
        if (!allowedCategories.has(category)) {
          violations.push({ file: filepath, category });
        }
      }
    }
  }

  return violations;
}

// Main execution
const allowed = readAllowedCategories();

if (allowed) {
  const inspectData = runQuartoInspect();
  const violations = validateCategories(inspectData, allowed);

  if (violations.length > 0) {
    console.error("\nCategory validation failed!\n");
    console.error("The following files use categories not listed in _qant.yml:\n");

    for (const v of violations) {
      console.error(`  ${v.file}: '${v.category}'`);
    }

    console.error("\nPlease add these categories to _qant.yml or remove them from the documents.\n");
    Deno.exit(1);
  } else {
    console.log("All categories validated successfully.");
  }
}

File: _extensions/qant/qant.lua
-- qant: Categorical imperative for Quarto categories
-- Validates document categories against a centralized allow-list

local function read_allowed_categories()
  -- Read _qant.yml to get allowed categories
  local file = io.open("_qant.yml", "r")
  if not file then
    quarto.log.output("Warning: _qant.yml not found. Skipping category validation.")
    return nil
  end

  local content = file:read("*all")
  file:close()

  -- Simple YAML parsing for categories list
  -- Look for "categories:" followed by list items
  local categories = {}
  for line in content:gmatch("[^\r\n]+") do
    local cat = line:match("^%s*-%s*(.+)$")
    if cat then
      categories[cat] = true
    end
  end

  return categories
end

local function run_quarto_inspect()
  -- Execute quarto inspect and capture output
  local handle = io.popen("quarto inspect")
  if not handle then
    error("Failed to run quarto inspect")
  end

  local json_output = handle:read("*all")
  handle:close()

  return quarto.json.decode(json_output)
end

local function validate_categories(inspect_data, allowed_categories)
  local violations = {}

  -- Iterate through all files in the inspect data
  for filepath, filedata in pairs(inspect_data.fileInformation) do
    if filedata.metadata and filedata.metadata.categories then
      local doc_categories = filedata.metadata.categories

      -- Check each category in the document
      for _, category in ipairs(doc_categories) do
        if not allowed_categories[category] then
          table.insert(violations, {
            file = filepath,
            category = category
          })
        end
      end
    end
  end

  return violations
end

-- Main execution
local allowed = read_allowed_categories()

if allowed then
  local inspect_data = run_quarto_inspect()
  local violations = validate_categories(inspect_data, allowed)

  if #violations > 0 then
    quarto.log.output("\nCategory validation failed!\n")
    quarto.log.output("The following files use categories not listed in _qant.yml:\n")

    for _, v in ipairs(violations) do
      quarto.log.output("  " .. v.file .. ": '" .. v.category .. "'")
    end

    quarto.log.output("\nPlease add these categories to _qant.yml or remove them from the documents.\n")
    error("Category validation failed")
  else
    quarto.log.output("All categories validated successfully.")
  end
end
